### PR TITLE
update metaDataTags handling, saving as a group of datasets

### PR DESCRIPTION
**Pull request by fangq**
hi Jay, I spent the afternoon reading your Snirf class source code, and made some changes to the `metaDataTags` handling.

The updated `metaDataTags` class now uses a struct to handle the key/value pairs. It also has a Set()/Get()  method.

Handling metaDataTags entirely as a struct instead of per-item read/write makes file access much faster. It also supports different data types per subitem.

### CODE REVIEW

**Comment by dboas** [comment]
@jayd1860 and @fangq , what is the status of this pull request? Is it relevant after the discussions this past week?

**Comment by fangq** [comment]
if Jay had already implemented his own struct-based metaDataTags, then, he can close this one.

otherwise, he can use my code by doing a manual merge. the patch is not that long, so should be quite easy to merge.

**Comment by jayd1860** [comment]
Homer3 is SNIRF compliant as far as I know but it would be good to test by someone other then me. I added the SNIRF files produced by Homer3 to fnirs/snirf-samples and sent a pull request.  If Qianqian's reader can load them successfully then Homer3 is snirf-compliant.

**Comment by jayd1860** [comment]
I can close this now but if Qianqian has a reader that can load successfully the new sample files in the pull request that would give me more confidence in Homer3 before closing this issue.

**Comment by fangq** [comment]
hi @jayd1860 and @dboas, I have accepted the sample data, and updated the sample submodule to snirf. you see the samples here

https://github.com/fNIRS/snirf

I confirm that I can load your sample files, I found two minor things (previously also bugged me when I tested my easyh5 toolbox):

1. please make your formatVersion to be 1.0, instead of 1.10. Our specification only defines 1.0 SNIRF format.

2. the strings variables I loaded from your sample files are all transposed - for example, for formatVersion: '1.10', it shows as
```
>> ss=loadsnirf('neuro_run01.snirf')
ss = 
    formatVersion: [5x1 char]
             nirs: [1x1 struct]
>> ss.formatVersion
ans =
1
.
1
0
```

this is because hdf5 uses row-major order and matlab uses column-major order. you will have to transpose the data before saving. 

https://github.com/fangq/easyh5/blob/master/saveh5.m#L236

you will have to `fliplr(size(data))` when creating the H5 space option

3. this is another annoyance I found with hdf5 - when saving datasets/groups,if you don't do anything, all subfields will be sorted in alphabetic order. this can be a problem if you prefer certain appearance order, say /nirs/data2 to appear before /nirs/data10, for example. To stored/load data in the "creation order", you need extra stuff - I figured it out for my toolbox, please see the patch in this issue I created 

https://github.com/fangq/easyh5/issues/1

- basically, when saving the data, you need to add the below options for a group

https://github.com/fangq/easyh5/commit/66de6e2388630ab4defb46d0024992533602d1bf#diff-4203ac0d8ad7bad17140a6aa360bd313R108-R109

when loading the data, you need to use link iterator (`H5L.iterate`) to read the data in creation order first

https://github.com/fangq/easyh5/commit/b09c80f03fbb6293f880827f9399f5fee8d4b34a#diff-4bbbb3f86d8c5c3f0a4328d9ea208796R86-R92

this is not critical, but just in case you are interested in adding such feature

4. another "good to have" feature is data compression. we mentioned about it in the spec, here is my patch to support compression in easyh5

https://github.com/fangq/easyh5/issues/2

**Comment by fangq** [comment]
@jayd1860, on the other hand, I uploaded an empty dataset (minimum snirf file) that I created using my functions

```
snf=snirfcreate
savesnirf(snf,'minimum_example.snirf')
```

I placed it in this folder
https://github.com/fNIRS/snirf-samples/tree/master/test

please let me know if you can read it

**Comment by jayd1860** [comment]
Qianqian, 

Thanks for the great feedback. I made the change of formatVersion to 1.0. Also I was able to load mininal test file successfully. 

Based on your minimal test file I made the defaults for metaDataTags to agree with it. 

I made a pull request for the new sample files. 

My thoughts regarding the difference in storage - row-major vs column-major - between matlab and HDF5 is that its kind of arbitrary which side an application compensates for this difference - the saving or the loading side. Homer3 SNIRF compensates for it when loading (now it should either row or column). It seems to me that it's up to each developer to know about this difference and to add code that accounts for it either way, since there's no universal standard order between various applications. 

Regarding subfields order being sorted by HDF5, that is annoying and I plan to deal with this and the compression  (yes would be a great to have) and performance issues after the fnirs course.

**Comment by dboas** [comment]
Qianqian,
I echo what Jay said, excellent feedback. Thanks

Jay,
It would be great to add those extra features QIanqian talked about.

Thanks,
David

**Comment by fangq** [comment]
@jayd1860, I updated the sample data.

I also noticed a small problem: in either of our sample files, the `metaDataTags.MeasurementDate` and Time did not agree with the format that I required in the latest snirf commit

https://github.com/fNIRS/snirf/blob/master/snirf_specification.md#nirsimetadatatags

so, I just updated my parser/sample files to use the ISO8601 format, see

https://github.com/fangq/jsnirf/commit/e52d47a36ff2cbdf4ede11161805015dc0d5442b
https://github.com/fNIRS/snirf-samples/commit/606e56e2394b162405a0192ddff5253892e1bb7e

feel free to update your sample data again. Otherwise, I am closing this ticket (the transpose issue - do it in saving or loading may be a future issue)

**Comment by jayd1860** [comment]
Ah ok. Good catch. I'll do the same.
